src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4433,14 +4433,30 @@ void C2_MacroAssembler::vector_maskall_operation(KRegister dst, Register src, in
   }
 }
 
-void C2_MacroAssembler::vbroadcastd(XMMRegister dst, int imm32, Register rtmp, int vec_enc) {
-  if (VM_Version::supports_avx512vl()) {
-    movl(rtmp, imm32);
-    evpbroadcastd(dst, rtmp, vec_enc);
+void C2_MacroAssembler::vbroadcast(BasicType bt, XMMRegister dst, int imm32, Register rtmp, int vec_enc) {
+  int lane_size = type2aelembytes(bt);
+  bool is_LP64 = LP64_ONLY(true) NOT_LP64(false);
+  if ((is_LP64 || lane_size < 8) &&
+      ((is_non_subword_integral_type(bt) && VM_Version::supports_avx512vl()) ||
+       (is_subword_type(bt) && VM_Version::supports_avx512bw()))) {
```

Comment:
The vl check is needed for all lane sizes when vector width < 64 bytes. The check doesn't seem to capture that.

Reply:
DONE